src/matUtils/convert.cpp

```diff
@@ -187,52 +188,83 @@ int *make_allele_codes(int8_t ref, std::map<int8_t, uint> &alts) {
               //fprintf(vcf_file, "%s\t%d\t%s\t%c\t%s\t.\t.\t%s",
                     //   chrom.c_str(), pos, id .c_str(), MAT::get_nuc(ref), alt_str.c_str(),
                     //   info.c_str());
-              vcf_file << boost::format("%s\t%d\t%s\t%c\t%s\t.\t.\t%s")
-                      % chrom.c_str() % pos % id .c_str() % MAT::get_nuc(ref) % alt_str.c_str() % info.c_str();
+              std::string* out=new std::string(boost::str(boost::format("%s\t%d\t%s\t%c\t%s\t.\t.\t%s")
+                      % chrom.c_str() % pos % id .c_str() % MAT::get_nuc(ref) % alt_str.c_str() % info.c_str()));
               if (print_genotypes) {
+                  out->reserve(leaf_count*2);
                   int *allele_codes = make_allele_codes(ref, alts);
                   //fprintf(vcf_file, "\tGT");
-                  vcf_file << "\tGT";
+                  out ->append("\tGT");
                   for (uint i = 0;  i < leaf_count;  i++) {
                       int8_t allele = gt_array[i];
-                      vcf_file << boost::format("\t%d") % allele_codes[allele];
+                      out->append("\t");
+                      out->append(std::to_string(allele_codes[allele]));
                       //fprintf(vcf_file, "\t%d", allele_codes[allele]);
                   }
               }
               //fputc('\n', vcf_file);
-              vcf_file << '\n';
+              out->append("\n");
+              return out;
+    }
+};
+struct Pos_Finder{
+    uint& pos;
+    const std::vector<int8_t *>& pos_genotypes;
+    uint operator()(tbb::flow_control& fc) const{
+        for (; pos<pos_genotypes.size(); pos++) {
+            if (pos_genotypes[pos]) {
+                pos++;
+                return pos-1;
             }
         }
+        fc.stop();
+        return -1;
+    }
+};
+
+void write_vcf_rows(std::ostream& vcf_file, MAT::Tree T, std::vector<MAT::Node*> &dfs, bool print_genotypes) {
+    // Fill in a matrix of genomic positions and sample genotypes in the same order as the
+    // sample names in the header, compute allele counts, and output VCF rows.
+    uint leaf_count = count_leaves(dfs);
+    // The int8_t here is mutation_annotated_tree.hpp's binary encoding of IUPAC nucleotide bases.
+    std::unordered_map<std::string, std::vector<int8_t *>> chrom_pos_genotypes;
+    std::unordered_map<std::string, std::vector<int8_t>> chrom_pos_ref;
+    std::vector<struct MAT::Mutation *> mut_stack;
+    r_add_genotypes(T.root, chrom_pos_genotypes, chrom_pos_ref, leaf_count, 0, mut_stack);
+    // Write row of VCF for each variant in chrom_pos_genotypes[chrom]
+    for (auto itr = chrom_pos_genotypes.begin();  itr != chrom_pos_genotypes.end();  ++itr) {
+        std::string chrom = itr->first;
+        std::vector<int8_t *> pos_genotypes = itr->second;
+        uint pos=0;
+        tbb::parallel_pipeline(80,tbb::make_filter<void,uint>(tbb::filter::serial_in_order,Pos_Finder{pos,pos_genotypes})&

```

Comment:
@yceh How have you determined this hard-coded value 80?

Reply:
It was arbitrary. Ideally it would be something slightly larger than thread count, but I am trying to figure out where can I get the threads count.

Reply:
Maybe use `tbb::task_scheduler_init::default_num_threads()`?

Reply:
Thanks a lot. It works.